node/metered-channel/src/oneshot.rs

```diff
@@ -0,0 +1,348 @@
+		let Self { inner, name: _, shared_meter: _ } = self;
+		inner.send((Instant::now(), t)).map_err(|(_, t)| t)
+	}
+
+	/// Poll if the thing is already cancelled.
+	pub fn poll_canceled(&mut self, ctx: &mut Context<'_>) -> Poll<()> {
+		self.inner.poll_canceled(ctx)
+	}
+
+	/// Access the cancellation object.
+	pub fn cancellation(&mut self) -> Cancellation<'_, (Instant, T)> {
+		self.inner.cancellation()
+	}
+
+	/// Check the cancellation state.
+	pub fn is_canceled(&self) -> bool {
+		self.inner.is_canceled()
+	}
+
+	/// Verify if the `receiver` is connected to the `sender` [`Self`].
+	pub fn is_connected_to(&self, receiver: &OneshotMeteredReceiver<T>) -> bool {
+		self.inner.is_connected_to(&receiver.inner)
+	}
+}
+
+/// Oneshot receiver, created by [`channel`].
+#[derive(Debug)]
+pub struct OneshotMeteredReceiver<T> {
+	name: &'static str,
+	inner: oneshot::Receiver<(Instant, T)>,
+	/// Soft timeout, on expire a warning is printed.
+	soft_timeout_fut: Option<Fuse<Delay>>,
+	soft_timeout: Duration,
+	/// Hard timeout, terminating the sender.
+	hard_timeout_fut: Option<Delay>,
+	hard_timeout: Duration,
+	/// The first time the receiver was polled.
+	first_poll_timestamp: Option<Instant>,
+	shared_meter: Meter,
+}
+
+impl<T> OneshotMeteredReceiver<T> {
+	pub fn close(&mut self) {
+		self.inner.close()
+	}
+
+	pub fn try_recv(&mut self) -> Result<Option<T>, Error> {
+		match self.inner.try_recv() {
+			Ok(Some((_when, val))) => Ok(Some(val)),
+			Err(e) => Err(Error::from(e)),
+			Ok(None) => Ok(None),
+		}
+	}
+}
+
+impl<T> OneshotMeteredReceiver<T> {
+	pub fn meter(&self) -> Meter {
+		self.shared_meter.clone()
+	}
+
+	fn update_meter(&mut self, sent_at_timestamp: Instant, reason: Reason) {
+		let delta: Duration = sent_at_timestamp.elapsed();
+		let val =
+			if delta.as_millis() > u64::MAX as u128 { u64::MAX } else { delta.as_millis() as u64 };
+		self.shared_meter.first_poll_till_end_in_millis.store(val, Ordering::Release);
+
+		self.shared_meter.reason.store(reason as u8, Ordering::Release);
+	}
+}
+
+impl<T> FusedFuture for OneshotMeteredReceiver<T> {
+	fn is_terminated(&self) -> bool {
+		self.inner.is_terminated()
+	}
+}
+
+impl<T> Future for OneshotMeteredReceiver<T> {
+	type Output = Result<T, Error>;
+
+	fn poll(mut self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Result<T, Error>> {
+		let mut me = self.as_mut();
+
+		if me.first_poll_timestamp.is_none() {
+			me.first_poll_timestamp = Some(Instant::now());
+			me.soft_timeout_fut = Some(Delay::new(me.soft_timeout.clone()).fuse());
+			me.hard_timeout_fut = Some(Delay::new(me.hard_timeout.clone()));
+		}
+
+		let soft_timeout = unsafe {
+			self.as_mut().map_unchecked_mut(|this| this.soft_timeout_fut.as_mut().unwrap())
+		};
+		if soft_timeout.poll(ctx).is_ready() {
+			tracing::warn!("Oneshot `{name}` exceeded the soft threshold", name = &self.name);
+		}
+
+		let hard_timeout = unsafe {
+			self.as_mut().map_unchecked_mut(|this| this.hard_timeout_fut.as_mut().unwrap())
+		};
```

Comment:
can we please avoid unsafe? I saw you using pin-project before force-pushing

Reply:
Well, there is not really a difference. We need the projection one way or another. `pin-project` just didn't yield any benefits once you wrap `Option<impl Future>` with `Pin`, but you are correct.

Reply:
Ok, so I totally missed `Option::as_pin_mut`. Will adapt.

Reply:
@ordian replaced the impl with an alternative, `unsafe` free version, without `pin-project`, have to double check if all invariants are held correctly.